### PR TITLE
refactor garbage collector into individual tasks

### PR DESCRIPTION
**Pull request by agavra**
I wasn't totally happy with the way I had implemented the tickers/statuses in #343 so I have a minor PR here to refactor this here. 

The change does a few minor things:

1. Introduces `GcTask` as a trait, and three implementations `ManifestGcTask`, `WalGcTask`, `CompactedGcTask`
2. Moves all methods specific to one resource type into corresponding implementation of `GcTask`
3. Moves a few minor things such as ticker creation into `DirGcStatus` for simplicity

Mostly an opportunity for me to level up my Rust skills, so feel free to reject it if you don't like it!

### CODE REVIEW

File: src/garbage_collector.rs
Comment (criccomini) [inline src/garbage_collector.rs:147]: This is kind of nitty, but I generally prefer passing in params directly rather than the whole object (so manifest_store, db_stats, and so on). More verbose, but gives invoking code more flexibility (and easier to unit test).

**Review by criccomini — APPROVED** [review]
This looks great. I read it over and just have one nit. Mind rebasing so we can merge?

File: src/garbage_collector.rs
Comment (agavra) [reply src/garbage_collector.rs]: Thanks for the Windsurf rec:
image
Perfect on the first try 🤩

File: src/garbage_collector.rs
Comment (rodesai) [inline src/garbage_collector.rs:489]: not part of your patch, but stood out since we changed the select! conditions - shouldn't we be using `tokio::select!` here? `crossbeam_channel::select!` is synchronous so this would block the tokio runtime thread.

File: src/garbage_collector.rs
Comment (hachikuji) [inline src/garbage_collector.rs:540]: Nice refactor! One idea might be to let this function return the current manifest, which could then be passed through to `collect`. I think each task has to read the current manifest anyway.

File: src/garbage_collector.rs
Comment (agavra) [reply src/garbage_collector.rs:489]: Below is the code that spwans the thread. Wouldn't spawning a new thread not block the tokio runtime thread? Not sure how all this works in rust. I would imagine the `tokio_handle.block_on` part would handle the synchronicity part, but I'm a bit beyond my depth here. Perhaps we should leave changing this for another PR, since that would also involve changing all the ticker logic.

```rs
        thread::spawn(move || {
            let orchestrator = GarbageCollectorOrchestrator {
                manifest_store,
                table_store,
                options,
                external_rx,
                db_stats,
            };
            tokio_handle.block_on(orchestrator.run());

            // !important: make sure that this is always the last thing that this
            // thread does, otherwise we risk notifying waiters and leaving this
            // thread around after shutdown
            if shutdown_tx.send(true).is_err() {
                error!("Could not send shutdown signal to threads blocked on await_shutdown");
            }
        });
```

Comment (agavra) [reply src/garbage_collector.rs:489]: OK ChatGPT confirms your suspicion, the tokio thread will be blocked on this. I'm not sure why we have `tokio_handle.block_on` if we don't want that?

File: src/garbage_collector.rs
Comment (agavra) [reply src/garbage_collector.rs:540]: I like the idea and I tried to make this change, but it seems like each of these needs something slightly different:

1. expiring removed checkpoints needs just the latest manifest
2. `WalGcTask` and `CompactedGcTask` need all the active manifests
3. `ManifestGcTask` needs all manifests (including inactive, in order to clean them up)

I think it is possible to issue fewer reads to S3 if we do further refactoring, but I think it ends up complicating things a bit and I have to pass a lot of references around.

Let's leave this as a potential follow-up refactor.

File: src/garbage_collector.rs
Comment (rodesai) [reply src/garbage_collector.rs:489]: what this is doing is spawning a thread, and having that thread block until `orchestrator.run` completes on the tokio runtime. `orchestrator.run` still runs on the tokio runtime, and so if it blocks then it will block the runtime's threads, which is a no-no. I don't mind switching it over to tokio select after we merge this.

Comment (agavra) [reply src/garbage_collector.rs:489]: https://github.com/slatedb/slatedb/issues/380

**Pull request merged**
